File: Data Structures & Algorithms/two-integer-sum-ii/submission-1.py
class Solution:
    def twoSum(self, numbers: List[int], target: int) -> List[int]:
        left = 0
        right = len(numbers) - 1

        while left < right:
            sum = numbers[left] + numbers[right]
            if sum == target:
                return [left+1, right+1]
            elif sum < target:
                left += 1
                while left < right and numbers[left] == numbers[left-1]:
                    left +=1
            else:
                right -=1
                while left < right and numbers[right] == numbers[right + 1]:
                    right -= 1

File: Data Structures & Algorithms/two-integer-sum/submission-1.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> freq = new HashMap<>();
        Map<Integer, Integer> index = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            freq.put(nums[i], freq.getOrDefault(nums[i], 0) + 1);
            index.put(nums[i], i);
        }
        for (int i = 0; i < nums.length; i++) {
            int remainder = target - nums[i];
            freq.put(nums[i], freq.get(nums[i]) - 1);
            if (freq.get(nums[i]) == 0) {
                // freq.remove(nums[i]);
                index.remove(nums[i]);
                // System.out.printf("Removed %d\n", nums[i]);
            }
            if (index.getOrDefault(remainder, -1) != -1) {
                return new int[] {i, index.get(remainder)};
            }
        }
        return new int[] {-1, -1};
    }
}
